src/lib.rs

```diff
@@ -248,31 +258,307 @@ impl<T, const N: usize> ArrayVec<T, { N }> {
         let len = self.len();
 
         // bounds checks
-        if index > self.len() {
+        if index > len {
             out_of_bounds!("try_insert", index, len);
         }
         if self.is_full() {
             return Err(CapacityError(item));
         }
 
         unsafe {
-            // The spot to put the new value
-            let p = self.as_mut_ptr().add(index);
-            // Shift everything over to make space. (Duplicating the
-            // `index`th element into two consecutive places.)
-            ptr::copy(p, p.offset(1), len - index);
-            // Write it in, overwriting the first copy of the `index`th
-            // element.
-            ptr::write(p, item);
-            // update the length
-            self.set_len(len + 1);
+            self.insert_unchecked(index, item, len);
         }
 
         Ok(())
     }
 
+    /// Insert an item into the vector, removing and returning its last
+    /// item if already full.
+    ///
+    /// # Panics
+    ///
+    /// The item cannot be inserted at an index greater than the
+    /// vector's length or greater or equal to the maximum capacity `N`.
+    ///
+    /// # Exmaples
+    ///
+    /// When the vector's not full, [`ArrayVec::force_insert`] acts like
+    /// [`ArrayVec::insert`]:
+    ///
+    /// ```rust
+    /// use const_arrayvec::ArrayVec;
+    /// let mut vector: ArrayVec<u8, 5> = ArrayVec::new();
+    ///
+    /// assert_eq!(vector.force_insert(0, 42), None);
+    /// assert_eq!(vector.force_insert(0, 24), None);
+    /// assert_eq!(&vector, [24, 42].as_ref());
+    /// ```
+    ///
+    /// But when the vector's full, we get a different result:
+    ///
+    /// ```rust
+    /// use const_arrayvec::ArrayVec;
+    /// let mut vector = ArrayVec::from([
+    ///     "He".to_owned(),
+    ///     "ya".to_owned(),
+    /// ]);
+    ///
+    /// let out = vector
+    ///     .force_insert(1, "llo".to_owned())
+    ///     .unwrap();
+    ///
+    /// assert_eq!(&out, "ya");
+    /// assert_eq!(&vector, ["He".to_owned(), "llo".to_owned()].as_ref());
+    /// ```
+    pub fn force_insert(&mut self, index: usize, item: T) -> Option<T> {
+        let len = self.len();
+
+        // bound checks
+        if index > len {
+            out_of_bounds!("force_insert", index, len);
+        }
+
+        let result;
```

Comment:
If the `if { ... } else { ... }` chain below needs to assign `result` in every branch, what about writing `let result = if { ... } else { ... }`?

Reply:
I originally tried to write it as `let result = if { ... } else { ... }` but the last item must be read before it is overwritten in the `if` branch, which means that in order to be able to directly write `let result = if { ... } else { ... }` a temporary variable must be created which seemed unnecessary when `result` can just be assigned directly.
At least that was my reasoning behind it, if you still prefer `let result = if { ... } else { ... }` I can change it to be that way.

Reply:
Ah that makes sense :+1: